File: ConsoleApplication1/Trie.cs
using System.Collections.Generic;
using System.Linq;

namespace VersionCSharp
{
    public class Trie
    {
        private Node[] _nodes = new Node[26];


        public bool Contain(string word)
        {
            Node prev = null;
            var wordLength = word.Length;
            for (int i = 0; i < wordLength; i++)
            {
                var letter = word[i];
                if (i == 0)
                {
                    var current = _nodes[letter - 'a'];
                    if (current == null)
                    {
                        return false;
                    }
                    prev = current;
                }
                else
                {
                    var currentNode = prev.NextValues.SingleOrDefault(t => t.Letter == letter);
                    if (currentNode == null)
                    {
                        return false;
                    }
                    if (i == wordLength - 1)
                    {
                        return currentNode.IsLeaf;
                    }
                    prev = currentNode;
                }
            }

            return false;
        }


        public void AddWordToTrie(string word)
        {
            var wordLength = word.Length;
            Node prevNode = null;

            for (int i = 0; i < wordLength; i++)
            {
                var letter = word[i];
                if (i == 0) // this is first letter, we need to
                {
                    prevNode = SetFirstNode(letter);
                }
                else
                {
                    var currentNode = prevNode.NextValues.SingleOrDefault(t => t.Letter == letter);
                    if (currentNode == null)
                    {
                        currentNode = new Node() { Letter = letter };
                        prevNode.NextValues.Add(currentNode);
                        if (i == wordLength - 1)
                        {
                            currentNode.IsLeaf = true;
                        }
                        prevNode = currentNode;
                    }
                    else
                    {
                        prevNode = currentNode;
                    }
                }
            }
        }

        private Node SetFirstNode(char letter)
        {
            var node = _nodes[letter - 'a'];
            if (node == null)
            {
                node = new Node() { Letter = letter };
                _nodes[letter - 'a'] = node;
            }
            return node;
        }

        public bool SubContain(string word)
        {
            Node prev = null;
            var wordLength = word.Length;
            for (int i = 0; i < wordLength; i++)
            {
                var letter = word[i];
                if (i == 0)
                {
                    var current = _nodes[letter - 'a'];
                    if (current == null)
                    {
                        return false;
                    }
                    prev = current;
                }
                else
                {
                    var currentNode = prev.NextValues.SingleOrDefault(t => t.Letter == letter);
                    if (currentNode == null)
                    {
                        return false;
                    }
                    prev = currentNode;
                }
            }

            return true;
        }

        class Node
        {
            public Node()
            {
                NextValues = new List<Node>();
                IsLeaf = false;
            }

            public List<Node> NextValues { get; set; }
            public bool IsLeaf { get; set; }
            public char Letter { get; set; }
        }

        
    }
}

File: AlienLanguage/VersionFSharp/Trie.fs
module AlienLanguage.Trie

type Node = 
        { Letter : char;
          isLeaf : bool;
          mutable NextValues : List<Node> }

type Trie() = 
    let nodes = Array.zeroCreate<Option<Node>>(26)

    let rec recursCheck(nextValues : List<Node>, tail : List<char>, subContain : bool): bool = 
        let nextLetter = tail.Head
        let elem = nextValues |> List.tryFind (fun x -> x.Letter = nextLetter)
        match elem with
        | None -> false
        | Some y -> if (y.isLeaf || subContain) && tail.Length = 1 then true else recursCheck(y.NextValues, tail.Tail, subContain)

    let checkNode(node : Node, tail : List<char>, subContain : bool) : bool = 
        let letter = tail.Head
        if node.Letter = letter then 
            match tail.Tail.IsEmpty with
                | true -> true
                | _ -> recursCheck(node.NextValues, tail.Tail, subContain)
        else false
    
    let nodeIndex(letter : char) = (int)letter - (int)'a'

    let containCore(word : List<char>, subContain : bool) : bool = 
        if word.IsEmpty then false else
        match nodes.[nodeIndex(word.Head)] with
        | None -> false
        | Some y -> checkNode(y, word, subContain)

    let rec addNextLetter(nodePrev : Node, tail : List<char>) : Unit = 
        if tail.IsEmpty = false then
            let nextLetter = tail.Head
            let elem = nodePrev.NextValues |> List.tryFind (fun x -> x.Letter = nextLetter)
            if elem.IsSome && tail.IsEmpty = false then addNextLetter(elem.Value, tail.Tail)
            else 
                let node = { Letter = nextLetter; isLeaf = tail.Length = 1; NextValues = [] }
                nodePrev.NextValues <- node :: nodePrev.NextValues
                if (tail.IsEmpty = false) then addNextLetter(node, tail.Tail)

    member this.contain(word : List<char>) : bool = containCore(word, false)

    member this.subContain(word : List<char>) : bool = containCore(word, true)
    
    member this.addWord(word : List<char>)  = 
        let currentNode = nodes.[nodeIndex(word.Head)]
        if currentNode.IsNone then
            let node = Some { Letter = word.Head; isLeaf = false; NextValues = []}
            nodes.[nodeIndex(word.Head)] <- node
            addNextLetter(node.Value, word.Tail)
        else 
            addNextLetter(currentNode.Value, word.Tail)
